From c0ff103a75052a2b899110e1905270a0b0caa501 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sun, 2 Feb 2020 19:52:04 +0100
Subject: [PATCH] drivers/rng: add Amlogic hardware RNG driver

Add support for the hardware random number generator of Amlogic SOCs.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 drivers/rng/Kconfig     |  8 ++++++
 drivers/rng/meson-rng.c | 54 +++++++++++++++++++++++++++++++++++++++++
 2 files changed, 62 insertions(+)
 create mode 100644 drivers/rng/meson-rng.c

diff --git a/drivers/rng/Kconfig b/drivers/rng/Kconfig
index c1aa43b823..edb6152bb9 100644
--- a/drivers/rng/Kconfig
+++ b/drivers/rng/Kconfig
@@ -8,6 +8,14 @@ config DM_RNG
 
 if DM_RNG
 
+config RNG_MESON
+	bool "Amlogic Meson Random Number Generator support"
+	depends on ARCH_MESON
+	default y
+	help
+	  Enable support for hardware random number generator
+	  of Amlogic Meson SoCs.
+
 config RNG_SANDBOX
 	bool "Sandbox random number generator"
 	depends on SANDBOX
diff --git a/drivers/rng/meson-rng.c b/drivers/rng/meson-rng.c
new file mode 100644
index 0000000000..650557de59
--- /dev/null
+++ b/drivers/rng/meson-rng.c
@@ -0,0 +1,54 @@
+struct stm32_rng_platdata {
+};
+
+static int stm32_rng_read(struct udevice *dev, void *data, size_t len)
+{
+	return 0;
+}
+
+static int stm32_rng_init(struct stm32_rng_platdata *pdata)
+{
+	return 0;
+}
+
+static int stm32_rng_cleanup(struct stm32_rng_platdata *pdata)
+{
+	return 0;
+}
+
+static int stm32_rng_probe(struct udevice *dev)
+{
+	return 0;
+}
+
+static int stm32_rng_remove(struct udevice *dev)
+{
+	return 0;
+}
+
+static int stm32_rng_ofdata_to_platdata(struct udevice *dev)
+{
+	return 0;
+}
+
+static const struct dm_rng_ops meson_rng_ops = {
+	.read = meson_rng_read,
+};
+
+static const struct udevice_id meson_rng_match[] = {
+	{
+		.compatible = "amlogic,meson-rng",
+	},
+	{},
+};
+
+U_BOOT_DRIVER(meson_rng) = {
+	.name = "meson-rng",
+	.id = UCLASS_RNG,
+	.of_match = meson_rng_match,
+	.ops = &meson_rng_ops,
+	.probe = meson_rng_probe,
+	.remove = meson_rng_remove,
+	.platdata_auto_alloc_size = sizeof(struct meson_rng_platdata),
+	.ofdata_to_platdata = meson_rng_ofdata_to_platdata,
+};
-- 
2.24.1

